File: examples/sputterDeposition/sputterDeposition.cpp
#include <geometries/psMakeTrench.hpp>
#include <psProcess.hpp>
#include <psUtil.hpp>

#include <models/psCSVFileProcess.hpp>

using namespace viennaps;

constexpr int D = 2;
using NumericType = double;

#include <psMaterials.hpp>
#include <psVelocityField.hpp>

void runDeposition(SmartPointer<CSVFileProcess<NumericType, D>> &depoModel,
                   SmartPointer<Domain<NumericType, D>> &domain,
                   util::Parameters &params) {
  Process<NumericType, D>(domain, depoModel, params.get("depositionTime"))
      .apply();
}

int main(int argc, char **argv) {

  Logger::setLogLevel(LogLevel::ERROR);
  omp_set_num_threads(1);

  // Parse the parameters
  util::Parameters params;
  if (argc > 1) {
    params.readConfigFile(argv[1]);
  } else {
    std::cout << "Usage: " << argv[0] << " <config file> [visualize]"
              << std::endl;
    return 1;
  }

  // Optional visualization
  if ((argc >= 3 && std::string(argv[2]) == "visualize")) {
    std::string cmd = std::string("python3 visualizeDomain.py ") + argv[1];
    int ret = std::system(cmd.c_str());
    if (ret != 0) {
      std::cerr << "Visualization script failed or not found." << std::endl;
    }
  }

  // geometry setup
  auto geometry = SmartPointer<Domain<NumericType, D>>::New();

  MakeTrench<NumericType, D>(geometry, params.get("gridDelta"),
                             params.get("xExtent"), params.get("yExtent"),
                             params.get("trenchWidth"),
                             params.get("trenchDepth"),
                             params.get("taperingAngle"), 0.0, /* baseHeight */
                             false,       /* periodicBoundary */
                             false,       /* makeMask */
                             Material::Si /* material */
                             )
      .apply();

  geometry->saveVolumeMesh("Trench");
  geometry->duplicateTopLevelSet(Material::SiO2);

  auto direction = Vec3D<NumericType>{0., -1.0, 0.};

  std::string ratesFile = params.get<std::string>("ratesFile");
  auto offset = Vec2D<NumericType>{0., 0.};
  offset[0] = params.get<NumericType>("offsetX");

  auto depoModel = SmartPointer<CSVFileProcess<NumericType, D>>::New(
      ratesFile, direction, offset);

  std::string interpModeStr = params.get<std::string>("interpolationMode");
  depoModel->setInterpolationMode(interpModeStr);
  if (interpModeStr == "custom") {
    depoModel->setCustomInterpolator([](const Vec3D<NumericType> &coord) {
      return 0.04 + 0.01 * std::sin(10.0 * coord[0]);
    });
  }

  const int numCycles = params.get<int>("numCycles");
  const std::string name = "TrenchDeposition_";

  int n = 0;
  geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");

  for (int i = 0; i < numCycles; ++i) {
    std::cout << "Cycle " << i + 1 << std::endl;
    runDeposition(depoModel, geometry, params);
    geometry->saveSurfaceMesh(name + std::to_string(n++) + ".vtp");
  }

  geometry->saveVolumeMesh("TrenchFinal");
}

File: examples/sputterDeposition/sputterDeposition.py
from argparse import ArgumentParser

# Argument parsing
parser = ArgumentParser(
    prog="trenchDeposition",
    description="Run trench deposition with a rate profile from a CSV file.",
)
parser.add_argument("-D", "-DIM", dest="dim", type=int, default=2)
parser.add_argument("filename")
parser.add_argument(
    "--visualize", action="store_true", help="Visualize the rate and geometry domains"
)
args = parser.parse_args()

import viennaps as ps

# Select dimension module
if args.dim == 2:
    print("Running 2D simulation.")
    import viennaps.d2 as psd
else:
    print("Running 3D simulation.")
    import viennaps.d3 as psd

# Setup
ps.Logger.setLogLevel(ps.LogLevel.ERROR)
ps.setNumThreads(16)

# Load parameters
params = ps.readConfigFile(args.filename)

if args.visualize:
    from visualizeDomain import visualize2d

    visualize2d(
        rates_file=params["ratesFile"].strip(),
        offset=params["offsetX"],
        x_extent=params["xExtent"],
        trench_width=params["trenchWidth"],
    )

# Geometry setup
geometry = psd.Domain()
psd.MakeTrench(
    domain=geometry,
    gridDelta=params["gridDelta"],
    xExtent=params["xExtent"],
    yExtent=params["yExtent"],
    trenchWidth=params["trenchWidth"],
    trenchDepth=params["trenchDepth"],
    taperingAngle=params["taperingAngle"],
    baseHeight=0.0,
    periodicBoundary=False,
    makeMask=False,
    material=ps.Material.Si,
).apply()

geometry.saveVolumeMesh("Trench")
geometry.duplicateTopLevelSet(ps.Material.SiO2)

# Setup direction
direction = [0.0, -1.0, 0.0]

# Setup offset
offset = [0.0, 0.0]
offset[0] = params["offsetX"]

# Create CSV-based deposition process
depoModel = psd.CSVFileProcess(
    ratesFile=params["ratesFile"].strip(),
    direction=direction,
    offset=offset,
)

# Select interpolation mode
mode = params["interpolationMode"].strip().lower()
depoModel.setInterpolationMode(mode)
if mode == "custom":

    def custom_interp(coord):
        import numpy as np

        x, _, _ = coord
        return 0.04 + 0.01 * np.sin(10.0 * x)

    depoModel.setCustomInterpolator(custom_interp)

# Simulation loop
numCycles = int(params["numCycles"])
filename_prefix = "TrenchDeposition_"

n = 0
geometry.saveSurfaceMesh(f"{filename_prefix}{n}.vtp")
n += 1

for i in range(numCycles):
    print(f"Cycle {i + 1}")
    psd.Process(geometry, depoModel, params["depositionTime"]).apply()
    geometry.saveSurfaceMesh(f"{filename_prefix}{n}.vtp")
    n += 1

geometry.saveVolumeMesh("TrenchFinal")
